otsserver/stamper.py

```diff
@@ -276,8 +276,8 @@ def __do_bitcoin(self):
                 # have been mined, and are waiting for confirmations.
                 self.txs_waiting_for_confirmation[block_height] = mined_tx
 
-                # Since all unconfirmed txs conflict with each other, we can clear the entire lot
-                self.unconfirmed_txs.clear()
```

Comment:
Actually we can't do this, because we remove the relevant commitments from pending at https://github.com/RCasatta/opentimestamps-server/blob/f2a4f5d91ff4f7163866287ab6cb741552ed2397/otsserver/stamper.py#L267. That merkle tree of operations is *not* saved - it's deterministically regenerated as needed by https://github.com/RCasatta/opentimestamps-server/blob/f2a4f5d91ff4f7163866287ab6cb741552ed2397/otsserver/stamper.py#L260 to save on memory - so you'll get them out of sync.

Note how even if you fix this every time someone timestamps one of our commitments for us we waste a transaction. We'll notice it and clear the `unconfirmed_txs` list, but we'll end up wasting a transaction as a brand new tx will be created from an unspent input as `find_unspent()` tries confirmed outputs first.

Maybe what we could do is have a per-unconfirmed_txs `pending_commitments` linked-list, with the list itself actually shared between subsequent unconfirmed txs unless one of the commitments was actually mined. Secondly, only actually clear `unconfirmed_txs` if a tx in that set was mined. However note that an attacker can *still* grief us by repeatedly mining old commitments tips - each time they do we'd make another copy of the pending commitments set in RAM.

Heck, the best approach is probably to just have a separate pending commitments set for every unconfirmed tx - at least then the memory usage is predictable! There's a setting for maximum number of pending commitments per tx, so just use that to keep RAM usage sufficiently low.

This would be much simpler with the original OpenTimestamps scheme of just having a giant MMR of all commitments ever made and timestamping that... :/

Reply:
sorry for not replying in thread...  I was on the phone.
Yes, MMR would have been great